Added `JointRV`  and `MarginalDistribution` in stats

akash9712 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->
Fixes #14852 Closes #14931 

#### Brief description of what is fixed or changed
Added `JointRV`, which allows the user to create a joint random variable
with each of its component varying over the real line. Also added
`JointDistributionHandmade` for this purpose. 
Added `MarginalDistribution` to allow random variables with components marginalized out. This also adds the ability to handle compound random variables by marginalizing over latent distributions, with or without a given condition.

<!-- BEGIN RELEASE NOTES -->
* stats
    * Add `JointRV` to allow user to create Joint random variables
    * Add `MarginalDistribution`
    * Ability to handle compound distributions 
<!-- END RELEASE NOTES -->

#### Other comments
Because the PR uses indexed symbols in `Lambda`, which raised an error, suitable modifications were made in core/function.py.

asmeurer committed:
Fix some 3-space indents

asmeurer committed:
init_printing: don't print builtin container subtypes that have custom printers

For example, sys.flags should print like

sys.flags(debug=0, py3k_warning=0, division_warning=0, division_new=0,
inspect=0, interactive=0, optimize=0, dont_write_bytecode=0, no_user_site=0,
no_site=0, ignore_environment=0, tabcheck=0, verbose=0, unicode=0,
bytes_warning=0, hash_randomization=0)

but before this it would print as

(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0)

with init_printing() used.

There is a test failure here that I'm not yet sure how to fix.

asmeurer committed:
Check for __str__ and __repr__ on the type, not the instance

asmeurer committed:
Fix builtin containers interactive test

asmeurer committed:
Add init_printing test for tuple that should pretty print

jlumpe committed:
Modify codegen.ast.Assignment class hierarchy.

Remove Relational from class hierarchy, also add common base class for both
Assignment and AugmentedAssignment so it is easier to differentiate
between them using isinstance().

jlumpe committed:
Add args test for AssignmentBase

jlumpe committed:
Update doctest for new ast.Assignment str printing

NikhilPappu committed:
Added a parser to parse Autolev code to SymPy code.

akash9712 committed:
[WIP] Add `JointDistributionHandmade`

Added `JointDistributionHandmade` and two end user functions to allow
users to create joint distributions with a given PDF.

eward-sunbelt committed:
Add dtype optional argument to __array__

eward-sunbelt committed:
Add test that matrix __array__ takes dtype argument

eward-sunbelt committed:
Update to test for dtype explicitly in np object

akash9712 committed:
Added `JointRV`

Added `JointRV`, which allows the user to create a joint random variable
with each of its component varying over the real line. Added
`JointDistributionHandmade` for this purpose. Also made minor
modification in core/function.py to allow `Lambda` to accept indexed
symbols. Added tests.

akash9712 commented:
Ping @Upabjojr @jksuom 
Can you please restart the travis build? The failures seem random.

jksuom commented:
The failures seem related to JointRV. For example,
```
______________________ sympy.stats.joint_rv_types.JointRV ______________________
File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/sympy-1.2.1.dev0-py2.7.egg/sympy/stats/joint_rv_types.py", line 48, in sympy.stats.joint_rv_types.JointRV
Failed example:
    density(N1)(1, 2)
Expected:
    exp(-2)/(2*pi)
Got:
    exp(-1)/(2*pi)
```
and
```
_______________ sympy/stats/tests/test_joint_rv.py:test_JointRV ________________
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.4.6/lib/python3.4/site-packages/sympy-1.2.1.dev0-py3.4.egg/sympy/utilities/runtests.py", line 1264, in _timeout
    function()
  File "/home/travis/virtualenv/python3.4.6/lib/python3.4/site-packages/sympy-1.2.1.dev0-py3.4.egg/sympy/stats/tests/test_joint_rv.py", line 71, in test_JointRV
    str(sqrt(2)*exp(-S(1)/2)*erf(sqrt(2)*x1/2)/(4*sqrt(pi)))
AssertionError
```

akash9712 committed:
Fix in `JointRV`

Fix in `JointRV` to order indexed symbols extracted by their index,
instead of using the set as it is. Also replaced symbols by indexed
symbols in `marginal_distribution`. `Indexed` are not used directly and
instead replace `Symbol` towards the end of the function because
`integrate` raised error in some cases when the variable of integration
was of type `Indexed`.

Commit:
Improve docstrings for sphinx docs

Commit:
Misc LaTeX fixes in combinatorial

akash9712 committed:
[WIP]Added `MarginalDistribution`

Added class `MarginalDistribution` and the ability to handle compound
distributions using this.

Commit:
Use plural module names in discrete

ramvenkat98 committed:
Added test that no files are executable

Created test_executable.py which tests that only files with shebangs have
an executable bit set. These two files violated this and had permissions
changed (to code number 644):
    sympy/utilities/tests/test_module_imports.py
    data/TeXmacs/progs/init-sympy.scm

ramvenkat98 committed:
Changed permissions of executable files correctly

Changed permissions of these two files to 644:
    sympy/utilities/tests/test_module_imports.py
    data/TeXmacs/progs/init-sympy.scm

ramvenkat98 committed:
Corrected test script and added to travis

Made test script executable and added shebang line
Made the error message for failing the test clearer
Added the file to the tests in travis

cbm755 committed:
RisingFactorial/FallingFactorial: fix high-precision floats

Instead of changing the evalf table, add appropriate entries to the
`MPMATH_TRANSLATIONS` table in lambdify.  The `MPMATH_TRANSLATIONS`
fix is taken from the Diofant project, with appreciation:
skirpichev/diofant/commit/4bcd9640f83ac0e050d38b5e4772547070f7ed09

The remaining tests and changes are my own.

- - - -

Revert some earlier evalf changes: I don't really understand this code
but it seemed to be limited to double precision.  Replace two broken
tests (str comparisions of Floats are generally fragile and also the
provided strings were not accurate).  Replace with 32-digit versions of
the same tests, compared relatively to computations from another CAS.

This also makes `lambdify(..., 'mpmath')` functional: test for that.

Commit:
Improve LaTeX equations

Commit:
Use cdots for addition and multiplication

akash9712 committed:
Modifications in tests in stats

Modified some tests, and added `evaluate` flag for marginal
distributions.

akash9712 committed:
Added test for density of compound distribution

Upabjojr commented on sympy/stats/joint_rv.py:
```diff
@@ -30,7 +32,7 @@ def __new__(cls, sym, dist):
     @property
     def set(self):
-        return self.domain.set
+        return self.distribution.set
```
why this?

akash9712 replied:
I think self.domain should return an object of `RandomDomain`, which is not yet implemented for `JointPSpace`.

jksuom committed:
Merge pull request #14618 from asmeurer/init_printing-builtin-subclasses

init_printing: don't print builtin container subtypes that have custom printers

asmeurer committed:
Merge pull request #14971 from cbm755/rf_ff_hiprec_fix

RisingFactorial/FallingFactorial: fix high-precision floats

asmeurer committed:
Print rationals correctly in the mpmath lambdify printer

This prevents nsolve from losing precision when the equation has rationals.

Fixes #14974.

asmeurer committed:
Merge pull request #14976 from asmeurer/lambdify-mpmath-rational

Print rationals correctly in the mpmath lambdify printer

asmeurer committed:
Merge pull request #14944 from ethankward/master

Add dtype to matrix __array__ method for numpy

asmeurer committed:
Merge pull request #14970 from ramvenkat98/14845_executable_test

Added test that no files are executable

jlumpe committed:
Fix Assignment code/pretty/repr printing and tests

asmeurer committed:
Define is_zero for im and re

Fixes #14978.

certik committed:
Merge pull request #14758 from NikhilPappu/autolev_parser

Autolev parser

akash9712 committed:
Corrections in test_mix.py

Corrections in import statements.

Commit:
Fix links in README.rst

Commit:
Correct references in convolution module

asmeurer committed:
Merge pull request #14979 from asmeurer/imre_is_zero

Define is_zero for im and re

cbm755 committed:
factor_terms: don't assume `as_coeff_Mul()[0]` is real

As far as I can tell, this code really wants to know if -1 can be
extracted from each term.  Well, we don't need to check `< 0` to do
that; use `.extract_multiplicative(-1)` instead.  It should still be
cheap because `.as_coeff_Mul()` is supposed to be a cheap operation.

Before:
```
>>> %timeit factor_terms(-x - Catalan*I - 1)
483 µs ± 7.54 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

>>> %timeit factor_terms(-x - Catalan*I - 1)
474 µs ± 2.66 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

After:
```
>>> %timeit factor_terms(-x - Catalan*I - 1)
492 µs ± 8.87 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

>>> %timeit factor_terms(-x - Catalan*I - 1)
503 µs ± 12.4 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Commit:
Improve documentation in convolution module

jksuom committed:
Merge pull request #14969 from sidhantnagpal/improve-sphinx-docs

Improve documentation for sphinx docs

smichr committed:
Merge pull request #14980 from cbm755/non_real_factor

factor_terms: don't assume `as_coeff_Mul()[0]` is real

asmeurer committed:
Merge pull request #14866 from jlumpe/change-ast-Assignment-base

Change ast assignment base

akash9712 committed:
Merge branch 'Joint_Handmade' of https://github.com/akash9712/sympy into Joint_Handmade

akash9712 committed:
Partially fixed failing tests in stats.

Fixed some of the previously failing tests in stats module, by
constructing marginal distribution with existing distributions
instead of PDFs.

akash9712 committed:
Fixed `domain` for `JointPSpace`

Fixed `domain` for JointPSpace and made some corrections to `pdf` in
`MarginalDistribution`.
